src/cargo/util/toml/mod.rs

```diff
@@ -1435,24 +1436,53 @@ fn to_real_manifest(
             .unwrap_or_else(|| semver::Version::new(0, 0, 0)),
         source_id,
     );
-    let summary = Summary::new(
-        pkgid,
-        deps,
-        &resolved_toml
-            .features
-            .as_ref()
-            .unwrap_or(&Default::default())
-            .iter()
-            .map(|(k, v)| {
-                (
-                    InternedString::new(k),
-                    v.iter().map(InternedString::from).collect(),
-                )
-            })
-            .collect(),
-        resolved_package.links.as_deref(),
-        rust_version.clone(),
-    )?;
+    let summary = {
+        let mut summary = Summary::new(
+            pkgid,
+            deps,
+            &resolved_toml
+                .features
+                .as_ref()
+                .unwrap_or(&Default::default())
+                .iter()
+                .map(|(k, v)| {
+                    (
+                        InternedString::new(k),
+                        v.iter().map(InternedString::from).collect(),
+                    )
+                })
+                .collect(),
+            resolved_package.links.as_deref(),
+            rust_version.clone(),
+        );
+        // editon2024 stops exposing implicit features, which will strip weak optional dependencies from `dependencies`,
+        // need to check whether `dep_name` is stripped as unused dependency
+        if let Err(ref mut err) = summary {
+            if let Some(missing_dep) = err.downcast_mut::<MissingDependencyError>() {
+                if missing_dep.weak_optional() {
+                    // dev-dependencies are not allowed to be optional
+                    let mut orig_deps = vec![
+                        original_toml.dependencies.as_ref(),
+                        original_toml.build_dependencies.as_ref(),
+                    ];
+                    for (_, platform) in original_toml.target.iter().flatten() {
+                        orig_deps.extend(vec![
+                            platform.dependencies.as_ref(),
+                            platform.build_dependencies.as_ref(),
+                        ]);
+                    }
+                    for deps in orig_deps {
+                        if let Some(deps) = deps {
+                            if deps.keys().any(|p| *p.as_str() == missing_dep.dep_name()) {
+                                missing_dep.set_unused_dependency(true);
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
To me, it is a bit odd to mutate an error. Could we extract this login to a function, which accepts the `original_toml` and the `err` as arguments and returns a refined error?

By doing that, `MissingDependencyError` doesn't need the field `unused_dependency`. We could do the logic complete outside `core::summary`.`Summary` is used by both index file parsing and manifest parsing, but the missing dep feature error doesn't make sense for index parsing at this moment. Therefore, `MissingDependencyError` will emit only one kind of error message.

Also, we could make all fields `pub` because the error is not intended to be modified.

Reply:
Thanks for your suggestions. 

All have been rebased into [fix: improve message for inactive weak optional feature with edition2024](https://github.com/rust-lang/cargo/pull/14026/commits/1306429464e8182d987cf4c4715a162b04e10df2)

Reply:
And the last commit has added the `missing_dep_diagnostic`,  but I shamless copy the code, may be it can be improved in latter PR.